runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -294,4 +439,74 @@ function M.get_line_diagnostics(bufnr, line_nr, opts, client_id)
   return diagnostic_vim_to_lsp(vim.diagnostic.get(bufnr, opts))
 end
 
+--- Clear diagnostics from pull based clients
+--- @private
+local function clear(bufnr)
+  for _, namespace in pairs(_client_pull_namespaces) do
+    vim.diagnostic.reset(namespace, bufnr)
+  end
+end
+
+--- autocmd ids for LspRequest handlers per buffer
+--- @private
+--- @type table<integer,integer>
+local _autocmd_ids = {}
+
+--- Disable pull diagnostics for a buffer
+--- @private
+local function disable(bufnr)
+  if not _autocmd_ids[bufnr] then
+    return
+  end
+  api.nvim_del_autocmd(_autocmd_ids[bufnr])
+  _autocmd_ids[bufnr] = nil
+  clear(bufnr)
+end
+
+--- Enable pull diagnostics for a buffer
+---@param bufnr (integer) Buffer handle, or 0 for current
+---@private
+function M._enable(bufnr)
+  bufnr = resolve_bufnr(bufnr)
+
+  if _autocmd_ids[bufnr] then
+    return
+  end
+
+  _autocmd_ids[bufnr] = api.nvim_create_autocmd('LspRequest', {
+    buffer = bufnr,
+    callback = function(opts)
+      if not opts.data.request then
+        return
+      end
+      if opts.data.request.type ~= 'complete' then
+        return
+      end
+      if opts.data.request.method ~= 'textDocument/didChange' then
+        return
+      end
+      refresh({ bufnr = bufnr, only_visible = true })
+    end,
+    group = augroup,
+  })
+
+  api.nvim_buf_attach(bufnr, true, {
```

Comment:
```suggestion
  api.nvim_buf_attach(bufnr, false, {
```

Full buffer contents are not needed/ used